xclim/indices/generic.py

```diff
@@ -147,97 +147,128 @@ def fitfunc(arr):
-      Name of the univariate distribution, such as beta, expon, genextreme, gamma, gumbel_r, lognorm, norm
-      (see scipy.stats).
-    mode : {'min', 'max}
-      Whether we are looking for a probability of exceedance (max) or a probability of non-exceedance (min).
+    p : xr.DataArray
+      Distribution parameters returned by the `fit` function. The array should have dimension `dparams` storing the
+      distribution parameters, and attribute `scipy_dist`, storing the name of the distribution.
+    q : Union[float, Sequence]
+      Quantile to compute, which must be between 0 and 1 inclusive.
 
     Returns
     -------
     xarray.DataArray
-      An array of values with a 1/t probability of exceedance (if mode=='max').
+      An array of parametric quantiles estimated from the distribution parameters.
+
+    Notes
+    -----
+    When all quantiles are above 0.5, the `isf` method is used instead of `ppf` because accuracy is sometimes better.
     """
-    t = np.atleast_1d(t)
+    q = np.atleast_1d(q)
 
     # Get the distribution
+    dist = p.attrs["scipy_dist"]
     dc = get_dist(dist)
 
-    # Fit the parameters of the distribution
-    p = fit(da, dist)
-
     # Create a lambda function to facilitate passing arguments to dask. There is probably a better way to do this.
-    if mode in ["max", "high"]:
+    if np.all(q > 0.5):
 
         def func(x):
-            return dc.isf(1.0 / t, *x)
+            return dc.isf(1 - q, *x)
 
-    elif mode in ["min", "low"]:
+    else:
 
         def func(x):
-            return dc.ppf(1.0 / t, *x)
-
-    else:
-        raise ValueError(f"Mode `{mode}` should be either 'max' or 'min'.")
+            return dc.ppf(q, *x)
 
     data = dask.array.apply_along_axis(func, p.get_axis_num("dparams"), p)
 
     # Create coordinate for the return periods
     coords = dict(p.coords.items())
     coords.pop("dparams")
-    coords["return_period"] = t
-
+    coords["quantile"] = q
     # Create dimensions
-    dims = list(p.dims)
-    dims.remove("dparams")
-    dims.insert(0, "return_period")
-
-    # TODO: add time and time_bnds coordinates (Low will work on this)
-    # time.attrs['climatology'] = 'climatology_bounds'
-    # coords['time'] =
-    # coords['climatology_bounds'] =
+    dims = [d if d != "dparams" else "quantile" for d in p.dims]
 
     out = xr.DataArray(data=data, coords=coords, dims=dims)
     out.attrs = p.attrs
-    out.attrs["standard_name"] = f"{dist} quantiles"
+    out.attrs["standard_name"] = f"{dist} quantile"
     out.attrs[
         "long_name"
-    ] = f"{dist} return period values for {getattr(da, 'standard_name', '')}"
+    ] = f"{dist} return period values for {p.attrs.get('standard_name', '')}"
     out.attrs["cell_methods"] = (
         out.attrs.get("cell_methods", "") + " dparams: ppf"
     ).strip()
-    out.attrs["units"] = da.attrs.get("units", "")
-    out.attrs["mode"] = mode
+    out.attrs["units"] = p.attrs["original_units"]
+
     out.attrs["history"] = (
```

Comment:
You might want to use `xclim.core.formatting.update_history`? It does the same but also adds a timestamp and xclim's version.

Reply:
done